dashboard/app/models/user.rb

```diff
@@ -770,8 +747,52 @@ def self.from_omniauth(auth, params)
     omniauth_user
   end
 
+  def self.initialize_new_oauth_user(user, auth, params)
+    user.provider = auth.provider
+    user.uid = auth.uid
+    user.name = name_from_omniauth auth.info.name
+    user.user_type = params['user_type'] || auth.info.user_type
+    # Store emails, except when using Clever
+    user.email = auth.info.email unless user.user_type == 'student' && OAUTH_PROVIDERS_UNTRUSTED_EMAIL.include?(auth.provider)
+
+    if OAUTH_PROVIDERS_UNTRUSTED_EMAIL.include?(auth.provider) && User.find_by_email_or_hashed_email(user.email)
+      user.email = user.email + '.oauthemailalreadytaken'
+    end
+
+    if auth.provider == :the_school_project
+      user.username = auth.extra.raw_info.nickname
+      user.user_type = auth.extra.raw_info.role
+      user.locale = auth.extra.raw_info.locale
+      user.school = auth.extra.raw_info.school.name
+    end
+
+    # treat clever admin types as teachers
+    if CLEVER_ADMIN_USER_TYPES.include? user.user_type
+      user.user_type = User::TYPE_TEACHER
+    end
+
+    # clever provides us these fields
+    if user.user_type == TYPE_TEACHER
+      user.age = 21
+    else
+      # As the omniauth spec (https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema) does not
+      # describe auth.info.dob, it may arrive in a variety of formats. Consequently, we let Rails
+      # handle any necessary conversion, setting birthday from auth.info.dob. The later
+      # shenanigans ensure that we store the user's age rather than birthday.
+      user.birthday = auth.info.dob
+      user_age = user.age
+      user.birthday = nil
+      user.age = user_age
+    end
+    user.gender = normalize_gender auth.info.gender
+  end
+
   def oauth?
-    OAUTH_PROVIDERS.include? provider
+    if provider == 'migrated'
```

Comment:
@islemaster and @madelynkasula this might conflict with some stuff you're working on, the `oauth?` method is used in a bunch of ways such that having *any* oauth option might not be the right switching condition. I'm looking into this as well.

Reply:
good point. i'm running into this now when updating the frontend for migrated oauth-only students to be able to update their email addresses, so i'm going to copy the logic you have here rather than rolling out something new. then we can refactor from there if necessary